File: dart/lib/src/bindings/eth_validate_record_v2_content.dart
import 'dart:typed_data';

import 'package:solana/solana.dart' hide RpcClient;

import '../constants/addresses.dart';
import '../constants/records.dart';
import '../domain/get_domain_address.dart';
import '../errors/sns_errors.dart';
import '../instructions/instruction_types.dart';
import '../instructions/validate_roa_ethereum_instruction.dart';

/// Parameters for Ethereum validation of record V2 content
class EthValidateRecordV2ContentParams {
  const EthValidateRecordV2ContentParams({
    required this.domain,
    required this.record,
    required this.owner,
    required this.payer,
    required this.signature,
    required this.expectedPubkey,
  });

  /// The .sol domain name
  final String domain;

  /// The record enum object
  final Record record;

  /// The owner of the record/domain
  final Ed25519HDPublicKey owner;

  /// The fee payer of the transaction
  final Ed25519HDPublicKey payer;

  /// Ethereum signature (65 bytes: r + s + v)
  final Uint8List signature;

  /// Expected Ethereum public key (64 bytes: uncompressed secp256k1)
  final Uint8List expectedPubkey;
}

/// Validates record V2 content with Ethereum signature
///
/// This function mirrors js/src/bindings/ethValidateRecordV2Content.ts
///
/// [params] - The parameters for Ethereum validation
///
/// Returns the Ethereum validation instruction
Future<TransactionInstruction> ethValidateRecordV2Content(
  EthValidateRecordV2ContentParams params,
) async {
  final domainResult = await getDomainAddress(
    GetDomainAddressParams(
      domain: '${params.record.value}.${params.domain}',
      record: RecordVersion.v2,
    ),
  );

  var parentAddress = domainResult.parentAddress;
  final isSub = domainResult.isSub;

  if (isSub) {
    final parentResult = await getDomainAddress(
      GetDomainAddressParams(domain: params.domain),
    );
    parentAddress = parentResult.domainAddress;
  }

  if (parentAddress == null) {
    throw InvalidParentError('Parent could not be found');
  }

  final instruction = ValidateRoaEthereumInstruction(
    validation: 2, // Validation.Ethereum = 2
    signature: params.signature,
    expectedPubkey: params.expectedPubkey,
  );

  instruction.setParams(
    programAddress: recordsProgramAddress,
    systemProgram: systemProgramAddress,
    splNameServiceProgram: nameProgramAddress,
    feePayer: params.payer.toBase58(),
    record: domainResult.domainAddress,
    domain: parentAddress,
    domainOwner: params.owner.toBase58(),
    centralState: centralStateDomainRecords,
  );

  return instruction.build();
}

File: js/src/bindings/ethValidateRecordV2Content.ts
import { Buffer } from "buffer";

import {
  SNS_RECORDS_ID,
  validateEthSignature,
  Validation,
} from "@bonfida/sns-records";
import { PublicKey } from "@solana/web3.js";

import { NAME_PROGRAM_ID } from "../constants";
import { InvalidParentError } from "../error";
import { Record, RecordVersion } from "../types/record";
import { getDomainKeySync } from "../utils/getDomainKeySync";
import { parseSupportedTld, SNS_TLD } from "../utils/tld";

/**
 * Builds the instruction to validate a V2 record's content using an Ethereum
 * signature, proving that the Ethereum address stored in the record is
 * controlled by the domain owner.
 *
 * @param domain The full domain name including TLD (e.g. `"mydomain.sns"`)
 * @param record The record type to validate
 * @param owner The owner of the domain
 * @param payer The fee payer of the transaction
 * @param signature The 64-byte Ethereum signature over the record content
 * @param expectedPubkey The 20-byte Ethereum public key expected to match the signature
 * @returns A {@link TransactionInstruction} that validates the Ethereum signature
 */
export const ethValidateRecordV2Content = (
  domain: string,
  record: Record,
  owner: PublicKey,
  payer: PublicKey,
  signature: Buffer,
  expectedPubkey: Buffer,
) => {
  // Only allows .sns domains
  parseSupportedTld(domain, [SNS_TLD]);

  let { pubkey, parent, isSub } = getDomainKeySync(
    `${record}.${domain}`,
    RecordVersion.V2,
  );

  if (isSub) {
    parent = getDomainKeySync(domain).pubkey;
  }

  if (!parent) {
    throw new InvalidParentError("Parent could not be found");
  }

  const ix = validateEthSignature(
    payer,
    pubkey,
    parent,
    owner,
    NAME_PROGRAM_ID,
    Validation.Ethereum,
    signature,
    expectedPubkey,
    SNS_RECORDS_ID,
  );
  return ix;
};
